packages/utils/src/eslint-utils/getParserServices.ts

```diff
@@ -1,39 +1,47 @@
 import type * as TSESLint from '../ts-eslint';
-import type { ParserServices } from '../ts-estree';
+import type {
+  ParserServices,
+  ParserServicesWithTypeInformation,
+} from '../ts-estree';
 
 const ERROR_MESSAGE =
   'You have used a rule which requires parserServices to be generated. You must therefore provide a value for the "parserOptions.project" property for @typescript-eslint/parser.';
 
+type GetParserServicesResult<T extends boolean> = T extends true
+  ? ParserServices
+  : ParserServicesWithTypeInformation;
+
 /**
  * Try to retrieve typescript parser service from context
  */
 function getParserServices<
   TMessageIds extends string,
   TOptions extends readonly unknown[],
+  TAllowWithoutFullTypeInformation extends boolean = false,
 >(
   context: Readonly<TSESLint.RuleContext<TMessageIds, TOptions>>,
-  allowWithoutFullTypeInformation = false,
-): ParserServices {
+  allowWithoutFullTypeInformation: TAllowWithoutFullTypeInformation = false as TAllowWithoutFullTypeInformation,
+): GetParserServicesResult<TAllowWithoutFullTypeInformation> {
   // backwards compatibility check
   // old versions of the parser would not return any parserServices unless parserOptions.project was set
```

Comment:
So what I was addressing in this PR specifically was `parserServices.hasFullTypeInformation`.
I guess technically i'm not addressing the issue you filed.

I think we should leave this comment in because it allows us to provide a nice error message in case the user misconfigured something.

I'll update the comment to something better.